python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -1286,27 +1290,22 @@ def update_replace_dict_and_conflicts(
 
     def map_asset_specs(self, fn: Callable[[AssetSpec], AssetSpec]) -> "AssetsDefinition":
         mapped_specs = []
+        all_deps_per_asset_key: Dict[AssetKey, Set[AssetDep]] = {}
         for spec in self.specs:
             mapped_spec = fn(spec)
             if mapped_spec.key != spec.key:
                 raise DagsterInvalidDefinitionError(
                     f"Asset key {spec.key.to_user_string()} was changed to "
                     f"{mapped_spec.key.to_user_string()}. Mapping function must not change keys."
                 )
-            if (
-                # check reference equality first for performance
-                mapped_spec.deps is not spec.deps and mapped_spec.deps != spec.deps
-            ):
-                raise DagsterInvalidDefinitionError(
-                    f"Asset deps {spec.deps} were changed to {mapped_spec.deps}. Mapping function "
-                    "must not change deps."
-                )
 
+            all_deps_per_asset_key[spec.key] = set(mapped_spec.deps)
             mapped_specs.append(mapped_spec)
 
-        return self.__class__.dagster_internal_init(
+        assets_def_new_specs = self.__class__.dagster_internal_init(
```

Comment:
I can take a pass at restructuring. The reason I did this way was to avoid re-building the asset dependency structure, but perhaps that's actually unavoidable anyway. I do so much dictionary munging anyway I'm not convinced I've actually saved any time complexity